### DISCUSSION TITLE
鸿蒙开发和调试 ArkUI 步骤

### DISCUSSION THREAD

**Discussion by xuanzebin** [NervJS/taro#14766] [鸿蒙 && OpenHarmony 适配小组]
## 背景
Taro 目前正在适配最新的 HarmonyOS Stage 应用模型，基于最新的 API 10 进行组件和 API 的适配。

Taro 源码项目分支：**feat/v3.7-ets**

框架支持：目前仅支持使用 `react` 来进行开发，后续待组件和 API 逐渐完善后再考虑支持 Vue 等其他框架。

## Taro 适配鸿蒙时 Taro 源码主要相关的 packge 的文件结构和作用
```
.
└── Taro
    ├── taro-platform-harmony
    │   └── src
    │       ├── apis
    │       ├── components
    │       ├── program
    │       ├── runtime-ets
    │       ├── runtime-framework
    │       └── index.ts
    └── taro-vite-runner
        └── src
            ├── harmony
            │   ├── postcss
            │   ├── asset.ts
            │   ├── babel.ts
            │   ├── config.ts
            │   ├── emit.ts
            │   ├── entry.ts
            │   ├── ets.ts
            │   ├── index.ts
            │   ├── page.ts
            │   ├── pipeline.ts
            │   └── style.ts
            └── index.harmony.ts
```
#### index.ts
Taro 鸿蒙适配平台的入口文件。

#### src/apis
Taro 以小程序 API 为基准，针对 OpenHarmony API 10 进行的适配。目前所有的 API 都存放在此，新建和修改 API 则只需关注此文件夹即可。

#### src/components
Taro 以小程序组件为基准的，利用 ArkTS 提供的自定义组件实现的一套组件 UI，包含常见的 View、Image、Text 等 UI 组件的实现。

#### src/program
Taro 针对鸿蒙平台提供的平台基类，用于编译的启动、配置修改等功能。

#### src/runtime-ets
Taro 针对鸿蒙平台实现的一套新的运行时方案，里面包含了在鸿蒙环境里可以跑通的 DOM、BOM 对象的实现。

#### src/runtime-framework
Taro 针对 React、Solid 等框架实现的与鸿蒙系统的桥接。

#### index.harmony.ts
Taro 鸿蒙平台编译器入口文件。

#### src/harmony/postcss
Taro 针对鸿蒙项目样式文件的处理。

#### src/harmony/asset.ts
Taro 针对鸿蒙项目依赖、图片和字体等资源的处理。

#### src/harmony/babel.ts
Taro 针对鸿蒙项目 ts 和 js 文件的转义处理。

#### src/harmony/config.ts
Taro 针对鸿蒙项目 VIte 编译配置生成。

#### src/harmony/emit.ts
Taro 针对鸿蒙项目编译产物抛出。

#### src/harmony/entry.ts
Taro 针对鸿蒙项目入口文件处理。

#### src/harmony/ets.ts
Taro 针对运行时生成的 ets 文件处理。

#### src/harmony/page.ts
Taro 针对鸿蒙项目的页面文件处理。

#### src/harmony/style.ts
调用 `src/harmony/postcss` 中的逻辑，样式处理插件的入口。

### 1. 申请成为华为合作伙伴，获得下载和使用 OpenHarmony 开发套件的白名单
（1）登录  [HarmonysOS 应用开发门户](https://link.juejin.cn/?target=https%3A%2F%2Fdeveloper.harmonyos.com%2Fcn%2Fhome)，点击右上角注册按钮，注册开发者帐号。

（2）进入  [HUAWEI DevEco Studio 套件货架中心](https://developer.harmonyos.com/deveco-developer-suite/)，申请白名单，由于目前最新版本的 OpenHarmony SDK 和 IDE 仍未对外开发，因此个人开发者若想尝鲜，需要先申请白名单成为合作伙伴后才能继续进行下面的步骤。

（3）白名单申请通过后，进入货架，下载 **IDE 版本为 DevEco Studio 4.0.3.600 的 DevEcoStudio4.0-API10** 开发套件。

（4）下载完成后，打开 IDE 安装包进行安装，安装成功后启动 DevEco Studio，根据[引导](https://developer.harmonyos.com/cn/docs/documentation/doc-guides-V4/sdk-prepar-0000001573170041-V4)解压套件里附带的 SDK 压缩包，并根据[引导](https://developer.harmonyos.com/cn/docs/documentation/doc-guides-V4/sdk-prepar-0000001573170041-V4)在 IDE 中配置好 SDK 的使用路径。

（5）SDK 配置成功后，看到设置面板中各个 SDK 的版本号与下面的图片中相同，则标识配置成功了。

![](https://img20.360buyimg.com/img/jfs/t1/185512/33/41206/63059/6544a44dFf833680e/4f8bbc23804e737a.png)

#### （1）创建 Harmony 项目
（a）创建新项目，选择需要开发的设备，然后 `Mode` 选择 `Stage` 模型，`Compile SDK` 选择 `4.0.0(API 10)`，按照引导操作后一个新的项目就被创建出来了。

（b）关注目录 `entry/src/main/ets/pages/Index.ets` 下面的文件，熟悉文件结构。`pages` 目录下为页面入口，新建项目的页面目录会包含若干个 `.ets` 文件，应用级配置信息位于 `build-profile.json5`，当前的模块信息 、编译信息配置项位于 `entry/build-profile.json5`。[项目结构详情](https://developer.harmonyos.com/cn/docs/documentation/doc-guides-V2/start-with-ets-stage-0000001558077129-V2)。

（c） 创建好项目后，根据[配置插件指引](https://developer.harmonyos.com/cn/docs/documentation/doc-releases-V2/import-plugin-0000001515948586-V2)，将下载的套件附带的插件配置在项目配置中。

（d） 修改项目的权限配置，根据项目需要在鸿蒙主项目 `entry/src/main/module.json5` 中所需要使用到的机器权限。
```json
{
  //...
  "requestPermissions": [
    {
      "name": "ohos.permission.VIBRATE"
    },
    {
      "name": "ohos.permission.GET_WIFI_INFO"
    },
    {
      "name": "ohos.permission.GET_NETWORK_INFO"
    },
    {
      "name": "ohos.permission.SET_NETWORK_INFO"
    },
    {
      "name": "ohos.permission.INTERNET"
    },
    {
      "name": "ohos.permission.GET_BUNDLE_INFO"
    },
    {
      "name": "ohos.permission.LOCATION"
    },
    {
      "name": "ohos.permission.APPROXIMATELY_LOCATION"
    },
    {
      "name": "ohos.permission.LOCATION_IN_BACKGROUND"
    }
  ]
}
```

#### （2）创建 Taro 项目
（a）若没有 Taro 脚手架，则需要提前安装一个最新版本的 Taro 脚手架
```bash
npm i -g @tarojs/cli
```
（b）使用命令 `taro init` 创建一个新的 Taro 项目，这里称这个项目为 Taro-Harmony-Demo，创建时需要注意，框架只能选择使用 `react` 的默认模板。

（c）将 Taro  项目中与 Taro 相关的、依赖版本和 Taro 脚手架版本相等的依赖筛选出来，将这些依赖的版本号调整至 `3.7.0-alpha.25`，重新安装依赖。

### 3. 拉取 Taro 最新源代码，并切换至鸿蒙开发分支 `feat/v3.7-ets`
（1） 拉取 Taro 仓库源代码并安装依赖
```bash
$ git clone git@github.com:NervJS/taro.git
```

（2）切换到鸿蒙开发分支，安装依赖并构建项目文件
```bash
$ git checkout feat/v3.7-ets
$ nvm use 16
$ pnpm i
$ pnpm run build
```

### 4. link 相关的 Taro package，并设置 Taro 项目 config 配置
（1）进入 Taro 源码仓库，到 `packages/taro-platform-harmony`、`packages/taro-vite-runner`、`packages/taro-service`、`packages/taro-react` 和 `packages/taro-plugin-react` 这五个 package 中执行 `yarn link`，再回到步骤 2 创建的 Taro 项目中，将这 5 个 package link 上去。

（2）之后进入到创建的 Taro-Harmony-Demo 项目中，调整 config/index.ts 的编译配置项
```js title="config/index.ts"
config = {
  // 配置使用插件
  plugins: ['@tarojs/plugin-platform-harmony'],
  // 将编译方式设置为使用 Vite 编译
  compiler: 'vite',
  // harmony 相关配置
  harmony: {
    // 【必填】鸿蒙主应用的绝对路径，例如：
    projectPath: path.resolve(process.cwd(), '../MyApplication'),
    // 【可选】HAP 的名称，默认为 'entry'
    hapName: 'entry',
    // 【可选】JS FA 的名称，默认为 'default'
    name: 'default',
  },
}
```

（3）设置 script 命令
```json
{
  // ...

  "scripts": {
    // ...
    "build:harmony": "taro build --type harmony",
    "dev:harmony": "npm run build:harmony -- --watch",
  },
}
```

### 5. 编译 Taro-Harmony-Demo 项目，运行和调试鸿蒙应用
（1）在 Taro-Harmony-Demo 项目根目录下执行 `yarn dev:harmony`，执行 Taro 项目在鸿蒙平台下的编译，编译后的文件会根据上一步的 projectPath 来进行输出。

（2）进入我们在步骤 2 中创建的鸿蒙项目，可以发现 Taro 编译的文件已经打到了该项目下，此时连接鸿蒙测试真机（需要联系华为的同事进行提供），即可在 IDE 进行对应的运行和调试 Debug。
![](https://img13.360buyimg.com/img/jfs/t1/226691/8/1877/144023/6544b5fdF3b303d52/e48a3b75e565036b.png)

以上就是开发和调试 Taro 适配鸿蒙 Stage 模型应用的流程和步骤，欢迎有共建想法的同学参与进来，并提交相关的 pr 和讨论～😆

**Comment by EchoHub** [comment]
构建鸿蒙应用后，在ide里面运行 提示找不到对应的npm包 Cannot find module '@tarojs/runtime' or its corresponding type declarations.

**Reply by xuanzebin** [reply]
4.0.0-beta.0 已发布，这个版本直接支持了鸿蒙适配的特性，可以直接安装这个版本的 taro-cli，然后按照文档：https://docs.taro.zone/docs/4.x/harmony 里面的开发指引进行开发。

**Reply by EchoHub** [reply]
构建出来的taro产出在devEco ide里面在构建时会提示 ArkTS:ERROR Failed to get an resolved OhmUrl by filepath "rollup_plugin_ignore_empty_module_placeholder"

**Reply by xuanzebin** [reply]
升级到 4.0.0-beta.1 版本试试

**Comment by LesterWeng** [comment]
请问无法使用 previewer 进行预览吗？预览时报错 
<img width="1365" alt="image" src="https://github.com/NervJS/taro/assets/23631443/7d6f13bb-f509-409c-8750-0a98154c90b0">

使用的是@tarojs/taro@4.0.0-canary.9，devEco build阶段没问题，仅previewer报错，我们暂时还没有测试机

**Reply by LesterWeng** [reply]
但使用devEco创建的hello word在previewer能正常预览

**Reply by xuanzebin** [reply]
首先，应该需要使用的 taro 版本是最新的 beta 版本，而非 canary 版本，目前的最新 beta 版本为 4.0.0-beta.8。
另外，previewer 只能预览和 EntryAbility 无关的页面，taro 构建的页面都会和 EntryAbility 有关联，因此无法使用 previewer 进行预览，如果没有测试机的话，推荐使用模拟器来进行运行。

**Reply by LesterWeng** [reply]
使用了最新的Taro v4.0.0-beta.23（所有@tarojs依赖），SDK是preview1(API11)，在模拟器运行，最简demo还是没能跑通，编译没有问题，模拟器空白，log不停抛出错误日志...
![image](https://github.com/NervJS/taro/assets/23631443/dc4dad19-68dd-420e-a2ae-ba03b576b8e9)

**Reply by xuanzebin** [reply]
执行 taro info 看看信息？

**Reply by xuanzebin** [reply]
或者提个 issues，提供更详细的报错信息或者鸿蒙工程 demo

**Reply by LesterWeng** [reply]
ok

**Reply by LesterWeng** [reply]
https://github.com/NervJS/taro/issues/15382
麻烦看下issue

**Comment by HeWenhui** [comment]
ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/image.ets:28:17
 Function return type inference is limited (arkts-no-implicit-return-types)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/image.ets:28:17
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/button.ets:17:20
 Argument of type 'ResourceColor | undefined' is not assignable to parameter of type 'ResourceColor'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/icon.ets:63:17
 Function return type inference is limited (arkts-no-implicit-return-types)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/icon.ets:63:17
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/video.ets:91:16
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:152:24
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:152:27
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:164:24
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:164:27
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:171:54
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:193:27
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:194:21
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:195:21
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:196:21
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:197:21
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:209:24
 Use explicit types instead of "any", "unknown" (arkts-no-any-unknown)

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/slider.ets:93:46
 Argument of type 'TaroSliderElement | null' is not assignable to parameter of type 'TaroElement<StandardProps<any, ITouchEvent>>'.
  Type 'null' is not assignable to type 'TaroElement<StandardProps<any, ITouchEvent>>'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/button.ets:110:4
 Property 'attributeModifier' does not exist on type 'ButtonAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/checkbox.ets:52:14
 Property 'attributeModifier' does not exist on type 'CheckboxAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/checkbox.ets:128:10
 Property 'attributeModifier' does not exist on type 'RowAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/checkbox.ets:137:10
 Property 'attributeModifier' does not exist on type 'ColumnAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/form.ets:17:6
 Property 'attributeModifier' does not exist on type 'RowAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/form.ets:30:6
 Property 'attributeModifier' does not exist on type 'ColumnAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/icon.ets:56:6
 Property 'attributeModifier' does not exist on type 'ImageAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/image.ets:27:6
 Property 'attributeModifier' does not exist on type 'ImageAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/input.ets:73:8
 Property 'attributeModifier' does not exist on type 'TextInputAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/label.ets:43:6
 Property 'attributeModifier' does not exist on type 'RowAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/label.ets:62:6
 Property 'attributeModifier' does not exist on type 'ColumnAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:149:14
 Property 'attributeModifier' does not exist on type 'TextPickerAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:161:14
 Property 'attributeModifier' does not exist on type 'TextPickerAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:190:14
 Property 'attributeModifier' does not exist on type 'DatePickerAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/picker.ets:206:14
 Property 'attributeModifier' does not exist on type 'TextPickerAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/radio.ets:56:14
 Property 'attributeModifier' does not exist on type 'RadioAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/radio.ets:131:10
 Property 'attributeModifier' does not exist on type 'RowAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/radio.ets:140:10
 Property 'attributeModifier' does not exist on type 'ColumnAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/richText.ets:16:6
 Property 'attributeModifier' does not exist on type 'RichTextAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/scrollView.ets:87:10
 Property 'attributeModifier' does not exist on type 'RowAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/scrollView.ets:100:10
 Property 'attributeModifier' does not exist on type 'ColumnAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/slider.ets:65:8
 Property 'attributeModifier' does not exist on type 'SliderAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/style.ets:13:36
 Cannot find name 'AttributeModifier'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/swiper.ets:48:4
 Property 'attributeModifier' does not exist on type 'SwiperAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/switch.ets:46:10
 Property 'attributeModifier' does not exist on type 'ToggleAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/text.ets:58:6
 Property 'attributeModifier' does not exist on type 'TextAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/textArea.ets:53:10
 Property 'attributeModifier' does not exist on type 'TextAreaAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/video.ets:84:6
 Property 'attributeModifier' does not exist on type 'VideoAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/view.ets:37:6
 Property 'attributeModifier' does not exist on type 'RowAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/view.ets:51:6
 Property 'attributeModifier' does not exist on type 'ColumnAttribute'.

ERROR: ArkTS:ERROR File: D:/DevEcoStudioProjects/Demo/entry/src/main/ets/npm/@tarojs/components/webview.ets:20:6
 Property 'attributeModifier' does not exist on type 'WebAttribute'.

COMPILE RESULT:FAIL {ERROR:50}
> hvigor ERROR: BUILD FAILED in 45 s 370 ms 
使用4.0版本构建后，编译报错，是因为我使用的devEco 是ArkTS语法严格校验版吗?

**Reply by xuanzebin** [reply]
目测是 DevEco Studio 的版本过于落后导致的，可以尝试将 DevEco Studio 版本升级至最新的 4.1.3.500，并替换对应的 SDK 试试

**Reply by Miaonster** [reply]
遇到了同样的问题，更新最新版解决，版本号 DevEco Studio NEXT Developer Preview1（Build Version: 4.1.3.501）

**Comment by HeWenhui** [comment]
用的是Build Version: 4.0.3.700 这个版本呢

**Reply by xuanzebin** [reply]
请替换至 4.1.3.500 试试，并保证 Taro 版本为最新的 beta 版本 4.0.0-beta.8

**Reply by HeWenhui** [reply]
我用的是新的4.0.3.700的版本，并也是用的taro版本4.0.0-beta.8. 鸿蒙API版本是10。我看taro 封装的的组件中，API底层没有对应的方法。例如TaroImage，调用底层Image组件，无attributeModifier方法。

**Reply by HeWenhui** [reply]
是我需要更新到最新的4.1.3.500的版本，API10的版本新增的方法了？

**Reply by xuanzebin** [reply]
API 11 新增的

**Comment by yFeii** [comment]
flex wrap 不支持了么

**Comment by raymond-cat** [comment]
@xuanzebin  有计划支持元服务吗，目前在鸿蒙api12上测试会提示部分属性和方法不支持元服务，希望taro4.0也能支持和兼容元服务
<img width="783" alt="image" src="https://github.com/NervJS/taro/assets/15725537/02479d6d-6bb6-462a-b31a-04509cad7823">

**Comment by sschris** [comment]
vue 的支持目前有计划了吗

**Comment by usernamedd** [comment]
安装了taro 4.1.1 ，目前看来应该是最新的，没有找到构建harmony应用的官方文档，看到的相关文档都是之前的。最明显的一个问题是，无论是taro build --type harmony  或者 taro build --type harmony_cpp (从taro文档找到的build type)，都会提示错误：Error: 不存在编译平台 harmony。
对harmony不支持了吗，或者是有新的配置方案与build type吗？
